### ISSUE TITLE
The choice of smart pointer depends on caller code instead of callee code.

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#3017]
I noticed that at https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/net.h#L73, there defines

```cpp
class PlainNet : public Net {
  void Run(const ScopePtr& scope,
           const platform::DeviceContext& dev_ctx) const override {
```

and at https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/net.h#L60, there defines

```cpp
class PlainNet : public Net {
  void InferShape(const ScopePtr& scope) const override {
```

where `ScopePtr` is defined at https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/scope.h#L27 as

```cpp
using ScopePtr = std::shared_ptr<Scope>;
```

However the choice of smart pointer, e.g., shared_ptr v.s. unique_ptr, depends on the application environment of class `Scope`. In some cases, we want to keep a singleton/default object and we use `unique_ptr<Scope>`; and in some other cases, we want auto garbage-collection and we use `shared_ptr<Scope>`. 

In above example, all what `PlainNet::InferShape` and `PlainNet::Run` want is `const Scope&`.

I'd suggest that we remove the assumption about what application code would want a smart pointer and keep the source code in its first-principle form.

**Comment by reyoung** [comment]
This argument is designed in purpose. Because we can only create `shared_ptr<Scope>` in Paddle because only `shared_ptr` of Scope can be `parent`.  If we use `unique_ptr<Scope>` or `const Scope&`, we cannot share that Scope when `Run` and `InferShape`.

**Comment by reyoung** [comment]
Fixed by #3027

**Comment by wangkuiyi** [comment]
I noticed that in PR #3027 , we made changes like
https://github.com/PaddlePaddle/Paddle/pull/3027/files#diff-fd81ea8da5bf0b61e95a74f8da7f4351L42, where  we changed 

```cpp
void AddOp(const OperatorPtr& o) = 0;
```

into 
```cpp
void AddOp(const std::shared_ptr<OperatorBase>& op) = 0;
```

I am afraid that such changes need to be reverted because

1. At testing time, the parameter to `AddOp` might not be allocated on the heap, but defined as a on-stack variable.  There would be no way to construct a smart pointer from it.

   ```cpp
   TEST(AddOp, AddOp) {
     MyOperator myop;
     Net mynet;
     mynet.AddOp(myop);
   ```

1. Locking reference count at every invocation to AddOp, and similar functions, would be too much costly.
